Shell/ConditionalRedundancyHandler.hpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+/**
+ * @file ConditionalRedundancyHandler.hpp
+ * Defines class ConditionalRedundancyHandler.
+ */
+
+#ifndef __ConditionalRedundancyHandler__
+#define __ConditionalRedundancyHandler__
+
+#include "Forwards.hpp"
+
+#include "Kernel/Ordering.hpp"
+
+#include "Lib/SharedSet.hpp"
+#include "Lib/Stack.hpp"
+
+#include "Inferences/DemodulationHelper.hpp"
+
+#include "Saturation/Splitter.hpp"
+
+#include "Options.hpp"
+
+namespace Shell {
+
+using namespace Lib;
+using namespace Indexing;
+
+using LiteralSet = SharedSet<Literal*>;
+
+struct ConditionalRedundancyEntry {
+  Term* lhs;
+  Term* rhs;
+  OrderingComparatorUP comp;
+  const LiteralSet* lits;
+  SplitSet* splits;
+  bool active = true;
+  unsigned refcnt = 1;
+
+  void deactivate() {
+    ASS(!splits->isEmpty());
+    active = false;
+  }
+
+  void obtain() {
+    refcnt++;
+  }
```

Comment:
Would it be worth wrapping this in e.g. a `std::shared_ptr` or at least RAIIng this?

Reply:
Code trees need a raw pointer, so at least there a shared pointer cannot be used (unless we heap allocate a shared pointer). Otherwise, what exactly you had in mind for RAIIing?

Reply:
Ah, the raw pointer thing makes sense then. The `obtain`/`release` stuff just looked like a `std::shared_ptr`.

Reply:
As an afterthought, if you think a heap-allocated `shared_ptr` is not weird, we can use that instead to get rid of this code.

Reply:
I don't think it's too weird, but let's keep it as-is for now. Later when I am feeling brave I can go looking for refcounting patterns and see if things can be improved.